firedrake/slate/slac/tsfc_driver.py

```diff
@@ -0,0 +1,98 @@
+from __future__ import absolute_import, print_function, division
+
+import collections
+
+from functools import partial
+
+from firedrake.slate.slate import Tensor
+from firedrake.slate.slac.utils import RemoveRestrictions
+
+from ufl.algorithms.map_integrands import map_integrand_dags
+from ufl import Form
+
+
+ContextKernel = collections.namedtuple("ContextKernel",
+                                       ["tensor",
+                                        "original_integral_type",
+                                        "tsfc_kernels"])
+
+
+def compile_terminal_form(tensor, tsfc_parameters=None):
+    """Compiles the TSFC form associated with a Slate :class:`Tensor`
+    object. This function will return a :namedtuple:`ContextKernel`
+    which stores information about the original tensor, integral types
+    and the corresponding TSFC kernels.
+
+    :arg tensor: A Slate `Tensor`.
+    :arg tsfc_parameters: An optional `dict` of parameters to provide
+                          TSFC.
+
+    Returns: A `ContextKernel` containing all relevant information.
+    """
+    from firedrake.tsfc_interface import compile_form as tsfc_compile
+
+    assert isinstance(tensor, Tensor), (
+        "Only terminal tensors have forms associated with them!"
+    )
+
+    mapper = RemoveRestrictions()
+    integrals = map(partial(map_integrand_dags, mapper),
+                    tensor.form.integrals())
+
+    transformed_integrals = transform_integrals(integrals)
```

Comment:
I assert that there can only be positive restrictions (if there are any restrictions at all). This is checked in `slate.py` by the `CheckRestrictions` multifunction. But I agree with revising the cell-facet info in the near future to allow for more general forms